privilege/privileges/cache.go

```diff
@@ -743,34 +767,43 @@ func decodeSetToPrivilege(s types.Set) mysql.PrivilegeType {
 	return ret
 }
 
-func (record *globalPrivRecord) match(user, host string) bool {
-	return record.User == user && patternMatch(host, record.patChars, record.patTypes)
+// hostMatch checks if giving IP is in IP range of hostname.
+// In MySQL, the hostname of user can be set to `<IPv4>/<netmask>`
+// e.g. `127.0.0.0/255.255.255.0` represent IP range from `127.0.0.1` to `127.0.0.255`,
+// only IP addresses that satisfy this condition range can be login with this user.
+// See https://dev.mysql.com/doc/refman/5.7/en/account-names.html
+func (record *baseRecord) hostMatch(s string) bool {
+	if record.hostIPNet == nil {
+		return false
+	}
+	ip := net.ParseIP(s).To4()
+	if ip == nil {
+		return false
+	}
+	return record.hostIPNet.Contains(ip)
 }
 
-func (record *UserRecord) match(user, host string) bool {
-	return record.User == user && patternMatch(host, record.patChars, record.patTypes)
+func (record *baseRecord) match(user, host string) bool {
```

Comment:
Because of above case, I think it's better to tell login match and verify match apart. When user with ip `127.0.0.1` login in, TiDB will match its ip with mask, it may get an auth hostname `127.0.0.0/255.0.0.0`. TiDB use this auth host name to verify user privilege, we don't need to do ip/mask match in verification phase.

Reply:
1 day later, I finally figure out this issue.

Reply:
To fix this bug, we need refactor amounts of code, I'd better to fix it in later PR. @imtbkcat